HDDS-11632. Publish images to GitHub container registry

adoroszlai opened the pull request:
## What changes were proposed in this pull request?

Currently `ozone-runner` images are built twice:

- by GitHub Actions workflow: used only for testing that the build works, discarded after CI run
- by Docker Hub automated build, when tags are pushed to the git repo

This PR changes the GitHub Actions workflow to publish the image to the GitHub Container Registry (but not for pull requests).

Images are versioned by commit SHA.  They are also tagged with git tag, if present.

Next steps planned:
- update `ozone-runner` image reference in Ozone CI
- replace Docker Hub automated build, pushing the image to Docker Hub from CI
- tweak build command to create multi-arch image

https://issues.apache.org/jira/browse/HDDS-11632

## How was this patch tested?

1. pushed commit: image was built ([workflow run](https://github.com/adoroszlai/ozone-docker-runner/actions/runs/11619655379), [image](https://github.com/users/adoroszlai/packages/container/ozone-runner/298599602?tag=522396802124849874e2b7b3ef8e02f65183107c))
2. pushed tag for same commit: build skipped, tag added to existing image ([workflow run](https://github.com/adoroszlai/ozone-docker-runner/actions/runs/11619794316), [image](https://github.com/users/adoroszlai/packages/container/ozone-runner/298599602?tag=tag1-HDDS-11632))
3. pushed new commit and new tag at once: image was built and tagged ([workflow run](https://github.com/adoroszlai/ozone-docker-runner/actions/runs/11619808734), [image](https://github.com/adoroszlai/ozone-docker-runner/pkgs/container/ozone-runner/298604288?tag=tag2-HDDS-11632))

Verified one the images:

```
$ docker pull ghcr.io/adoroszlai/ozone-runner:b57ad699321ea3b464e349c8b5cb215cabfdae43
...
Digest: sha256:ef2ac1e656f2487479efdf5c0b61e31b2d7f4fc50010e9c43d7441edb4d071e2
Status: Downloaded newer image for ghcr.io/adoroszlai/ozone-runner:b57ad699321ea3b464e349c8b5cb215cabfdae43
ghcr.io/adoroszlai/ozone-runner:b57ad699321ea3b464e349c8b5cb215cabfdae43

$ docker run -it --rm ghcr.io/adoroszlai/ozone-runner:b57ad699321ea3b464e349c8b5cb215cabfdae43 bash
To use Ozone please mount ozone folder to /opt/hadoop
[hadoop@24b37b308741 ~]$ 

$ docker pull ghcr.io/adoroszlai/ozone-runner:tag2-HDDS-11632                                      
tag2-HDDS-11632: Pulling from adoroszlai/ozone-runner
Digest: sha256:ef2ac1e656f2487479efdf5c0b61e31b2d7f4fc50010e9c43d7441edb4d071e2
Status: Downloaded newer image for ghcr.io/adoroszlai/ozone-runner:tag2-HDDS-11632
ghcr.io/adoroszlai/ozone-runner:tag2-HDDS-11632
```

adoroszlai commented:
Updated the patch:
- split the workflow into `build` and `tag`
- use `docker` actions instead of CLI where available
- avoid concurrent builds of the same image

This simplifies further changes to be done for building multi-arch images and pushing to multiple registries.

errose28 commented on .github/workflows/push.yaml:
```diff
@@ -0,0 +1,58 @@
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@9780b0c442fbb1117ed29e0efdff1e18412f7567
+        with:
+          registry: ghcr.io
```
We have the registry as a variable in build.yaml. It seems like we should either use a variable or hard code it in both places for consistency.

smengcl replied:
+1 Ethan. Should this be `${{ env.DEFAULT_REGISTRY }}` ?

adoroszlai replied:
`push.yaml` (then `tag.yaml`) defined the `DEFAULT_REGISTRY` variable initially, with a comment that it must be kept in sync between the two workflows.  But usage in `build.yaml` and here is different:

- `build.yaml` publishes to a single registry.
- `push.yaml` will eventually publish to multiple registries.  From this workflow's point of view, none of those are special.  The source image is taken as an input from the other workflow, so it doesn't need to know which registry is the default.

So I removed it.

On third thought I think we should hard-code it everywhere.

Here's why: `ghcr.io` is available for all forks, so the workflow works for everyone.  On the other hand, Docker Hub is special in that one needs to register and create the repo for these images.  Also, the secret for Docker Hub login must be set explicitly in the repo settings.  So login and push to Docker Hub would not work out-of-the-box for someone who just forks `apache/ozone-docker-runner`.  By the same logic, changing `DEFAULT_REGISTRY` to Docker Hub would break the workflow for such users.

errose28 reviewed (commented):
Thanks for working on this @adoroszlai. Overall looks good just a few minor questions/comments.

errose28 commented:
cc @ptlrs if you would like to review

adoroszlai commented on .github/workflows/build.yaml:
```diff
@@ -12,16 +12,75 @@
 name: build
+
+# This workflow builds the Docker image if it does not exists already.
+# For non-PR runs, it also publishes the image to the registry, tagging it by the full SHA of the commit.
```
I think we will mostly use `ozone-runner` by tag as previously (e.g. `20241104-jdk17-1`), so using shorter hash didn't seem necessary.

adoroszlai commented on .github/workflows/build.yaml:
```diff
@@ -12,16 +12,75 @@
-        run: DOCKER_BUILDKIT=1 docker build -t ghcr.io/$(echo $GITHUB_REPOSITORY | tr '[:upper:]' '[:lower:]' | sed 's/docker-//g') .
+      - name: Generate image ID
+        id: meta
+        uses: docker/metadata-action@8e5442c4ef9f78752691e2d8f8d19755c6f78e81
```
It works, but Apache Infra [policy](https://infra.apache.org/github-actions-policy.html) says we must use hashes.

errose28 replied:
Oh thanks for sharing, I wasn't aware of this. Looks like it only applies to external actions which I haven't had to use for apache workflows yet.

smengcl replied:
Thanks. I didn't know this before either. Might be for security concerns

adoroszlai commented:
Thanks @errose28, @smengcl for the reviews.  I have removed the `DEFAULT_REGISTRY` variable.  Also renamed the `push` workflow to `build-and-tag` to better reflect its purpose.

adoroszlai commented:
@errose28 @smengcl can you please take another look?

smengcl reviewed (approved):
lgtm. Thanks @adoroszlai

adoroszlai commented:
Thanks @errose28, @smengcl for the review.
